File: src/main/scala/com/vc/easy/L538.scala
package com.vc.easy

class TreeNode538(var _value: Int) {
  var value: Int = _value
  var left: TreeNode538 = null
  var right: TreeNode538 = null
}

object Solution {
  def convertBST(root: TreeNode538): TreeNode538 = {
    /**
    var sum = 0
    solve(root)
    def solve(root: TreeNode538): Unit = {
      if(root != null) {
        solve(root.right)
        val right = root.value
        root.value += sum
        sum += right
        solve(root.left)
      }
    }
    root*/
    var sum = 0
    import scala.collection.mutable
    val st = new mutable.Stack[TreeNode538]()
    var current = root
    while(current != null || st.nonEmpty) {
      if(current != null) {
        st.push(current)
        current = current.right
      }
      else {
        current = st.pop
        val oldValue = current.value
        current.value += sum
        sum += oldValue
        current = current.left
      }
    }
    root
  }
}

File: src/main/java/com/vc/easy/L538.java
package com.vc.easy;

import com.vc.hard.TreeNode;

class L538 {
    int res = 0;
    public TreeNode convertBST(TreeNode root) {
        helper(root);
        return root;
    }

    private void helper(TreeNode root) {
        if(root == null) return;
        helper(root.right);
        root.val += res;
        res = root.val;
        helper(root.left);
    }
}
